File: TriCoachAI/Core/PlanEngine/MesocycleBuilder.swift
import Foundation

/// Splits a macrocycle (phase) into 3-4 week mesocycle blocks, each with a
/// training focus. Blocks rotate the focus so a long Build phase progresses
/// aerobic threshold work into VO2max work rather than repeating itself.
enum MesocycleBuilder {
    private static func focuses(for phase: MacrocyclePhase) -> [String] {
        switch phase {
        case .base: return ["Endurance aérobie", "Endurance aérobie + technique"]
        case .build: return ["Seuil", "VO2max"]
        case .peak: return ["Spécifique course & enchaînements", "Allure course"]
        case .taper: return ["Affûtage"]
        case .transition: return ["Récupération active"]
        }
    }

    static func build(macrocycle: Macrocycle) -> [Mesocycle] {
        let calendar = Calendar.current
        let totalDays = calendar.dateComponents([.day], from: macrocycle.startDate, to: macrocycle.endDate).day.map { $0 + 1 } ?? 7
        let totalWeeks = max(1, Int((Double(totalDays) / 7.0).rounded(.up)))
        let blockSizeWeeks = 4
        let blockCount = max(1, Int((Double(totalWeeks) / Double(blockSizeWeeks)).rounded(.up)))
        let focusOptions = focuses(for: macrocycle.phase)

        var mesocycles: [Mesocycle] = []
        var cursor = macrocycle.startDate
        var weeksRemaining = totalWeeks

        for blockIndex in 0..<blockCount {
            let weeksInBlock = min(blockSizeWeeks, weeksRemaining)
            guard weeksInBlock > 0 else { break }
            let blockEnd = calendar.date(byAdding: .day, value: weeksInBlock * 7 - 1, to: cursor) ?? cursor
            let focus = focusOptions[blockIndex % focusOptions.count]

            mesocycles.append(
                Mesocycle(
                    name: "\(macrocycle.name) — bloc \(blockIndex + 1)",
                    focus: focus,
                    startDate: cursor,
                    endDate: min(blockEnd, macrocycle.endDate)
                )
            )
            cursor = calendar.date(byAdding: .day, value: weeksInBlock * 7, to: cursor) ?? cursor
            weeksRemaining -= weeksInBlock
        }
        return mesocycles
    }
}

File: tricoach-backend/src/modules/plans/engine/mesocycleBuilder.ts
import { randomUUID } from 'crypto';
import { addDays } from './periodizationEngine';
import { Macrocycle, MacrocyclePhase, Mesocycle } from './types';

const FOCUS_OPTIONS: Record<MacrocyclePhase, string[]> = {
  base: ['Endurance aérobie', 'Endurance aérobie + technique'],
  build: ['Seuil', 'VO2max'],
  peak: ['Spécifique course & enchaînements', 'Allure course'],
  taper: ['Affûtage'],
  transition: ['Récupération active'],
};

/** Direct port of Core/PlanEngine/MesocycleBuilder.swift. */
export const MesocycleBuilder = {
  build(macrocycle: Macrocycle): Mesocycle[] {
    const totalDays = Math.round((macrocycle.endDate.getTime() - macrocycle.startDate.getTime()) / 86400000) + 1;
    const totalWeeks = Math.max(1, Math.ceil(totalDays / 7));
    const blockSizeWeeks = 4;
    const blockCount = Math.max(1, Math.ceil(totalWeeks / blockSizeWeeks));
    const focusOptions = FOCUS_OPTIONS[macrocycle.phase];

    const mesocycles: Mesocycle[] = [];
    let cursor = new Date(macrocycle.startDate);
    let weeksRemaining = totalWeeks;

    for (let blockIndex = 0; blockIndex < blockCount; blockIndex++) {
      const weeksInBlock = Math.min(blockSizeWeeks, weeksRemaining);
      if (weeksInBlock <= 0) break;
      const blockEnd = addDays(cursor, weeksInBlock * 7 - 1);
      const focus = focusOptions[blockIndex % focusOptions.length]!;

      mesocycles.push({
        id: randomUUID(),
        name: `${macrocycle.name} — bloc ${blockIndex + 1}`,
        focus,
        startDate: cursor,
        endDate: blockEnd < macrocycle.endDate ? blockEnd : macrocycle.endDate,
        microcycles: [],
      });
      cursor = addDays(cursor, weeksInBlock * 7);
      weeksRemaining -= weeksInBlock;
    }
    return mesocycles;
  },
};
